File: 1694-MakeSumDivisibleByP/1694-MakeSumDivisibleByP.py
# Last updated: 10/1/2025, 12:47:16 AM
class Solution:
    def minSubarray(self, nums: List[int], p: int) -> int:
        n = len(nums)
        
        totalSum=0
        
        for num in nums:
            totalSum = (totalSum+num) % p
            
        if totalSum==0:
            return 0
        
        prefixModMap = {}
        
        prefixModMap[0]=-1
        
        prefixSum = 0
        
        minLen = n
        
        for i in range(0,n):
            prefixSum = (prefixSum+nums[i])%p
            target = (prefixSum-totalSum+p)%p
            
            if target in prefixModMap:
                minLen=min(minLen,i-prefixModMap[target])
                
            prefixModMap[prefixSum]=i
            
        if minLen == n:
            return -1
        else:
            return minLen
  

File: 1694-MakeSumDivisibleByP/1694-MakeSumDivisibleByP.cpp
// Last updated: 10/1/2025, 12:47:22 AM
class Solution {
public:
    int minSubarray(vector<int>& nums, int p) {
        int n = nums.size();

    int totalSum = 0;


    for (int num : nums) {

        totalSum = (totalSum + num) % p;

    }



    if (totalSum == 0) {

        return 0;

    }

    
    unordered_map<int, int> prefixModMap;

    prefixModMap[0] = -1;  
    int prefixSum = 0;

    int minLen = n;

    for (int i = 0; i < n; ++i) {

        

        prefixSum = (prefixSum + nums[i]) % p;

        

        int target = (prefixSum - totalSum + p) % p;

        

        if (prefixModMap.find(target) != prefixModMap.end()) {

            minLen = min(minLen, i - prefixModMap[target]);

        }

        

        prefixModMap[prefixSum] = i;

    }

    

    return minLen == n ? -1 : minLen;

}
    
};